src/charm.py

```diff
@@ -418,6 +446,52 @@ def _on_tls_certificate_pki_certificate_available(self, event: CertificateAvaila
         """Handle the tls-certificates-pki certificate available event."""
         self._add_ca_certificate_to_pki_secrets_engine()
 
+    def _on_vault_pki_certificate_creation_request(
+        self, event: CertificateCreationRequestEvent
+    ) -> None:
+        """Handle the vault-pki certificate creation request event."""
+        self._generate_pki_certificate_for_requirer(
+            event.certificate_signing_request, event.relation_id
+        )
+
+    def _generate_pki_certificate_for_requirer(self, csr: str, relation_id: int):
+        """Generate a PKI certificate for a TLS requirer."""
+        if not self.unit.is_leader():
+            logger.debug("Only leader unit can handle a vault-pki certificate request")
+            return
+        if not self._is_peer_relation_created():
+            logger.debug("Peer relation not created")
+            return
```

Comment:
Is it really needed to check the peer relation here?

Reply:
Yes, we need to get info from the peer relation to initialize the Vault client

Reply:
Do you mean `_get_initialization_secret_from_peer_relation` in `_get_initialized_vault_client`? I feel the name is a bit misleading, I know that we check for the peer relation when creating that secret and we put its id into relation data, but when getting the secret I don't think we really rely in the peer relation.
And we might get into a situation where the peer relation is created but the secret isn't. Maybe we should check for `_initialization_secret_set_in_peer_relation` instead of the peer relation itself.

Reply:
But in both cases, if we can't get the secret `_get_initialized_vault_client` will return `None`.

Reply:
It's not critical to remove it right now.

Reply:
This is true, the peer relation is no longer required for this. But like you said it's not critical here and we can address it later. That's why I've ignored it.